tests/hierarchy_driver.cc

```diff
@@ -1,11 +1,88 @@
 #include <mfmg/common/exceptions.hpp>
 
+#include <deal.II/base/timer.h>
+
 #include <boost/program_options.hpp>
 
 #include "test_hierarchy_helpers.hpp"
 
+template <int dim, int fe_degree>
+void matrix_free_two_grids(std::shared_ptr<boost::property_tree::ptree> params)
+{
+
+  using DVector = dealii::LinearAlgebra::distributed::Vector<double>;
+
+  MPI_Comm comm = MPI_COMM_WORLD;
+
+  dealii::ConditionalOStream pcout(
+      std::cout, dealii::Utilities::MPI::this_mpi_process(comm) == 0);
+
+  // Print a table with the timings when the destructor is called
+  auto timer = std::make_shared<dealii::TimerOutput>(
+      comm, pcout, dealii::TimerOutput::summary,
+      dealii::TimerOutput::wall_times);
+
+  auto material_property =
+      MaterialPropertyFactory<dim>::create_material_property(
+          params->get<std::string>("material_property.type"));
+  Source<dim> source;
+
+  auto laplace_ptree = params->get_child("laplace");
+  LaplaceMatrixFree<dim, fe_degree, double> mf_laplace(comm);
+  mf_laplace.setup_system(laplace_ptree, *material_property);
+
+  auto const locally_owned_dofs = mf_laplace._locally_owned_dofs;
+  DVector solution(locally_owned_dofs, comm);
+  DVector rhs(mf_laplace._system_rhs);
+
+  std::default_random_engine generator;
+  std::uniform_real_distribution<typename DVector::value_type> distribution(0.,
+                                                                            1.);
+  for (auto const index : locally_owned_dofs)
+  {
+    // Make the solution satisfy the Dirichlet conditions because these should
+    // be treated outside the preconditioner
+    if (mf_laplace._constraints.is_constrained(index))
+      solution[index] = 0.;
+    else
+      solution[index] = distribution(generator);
+  }
+
+  auto evaluator =
+      std::make_shared<TestMFMeshEvaluator<dim, fe_degree, double>>(
+          mf_laplace._dof_handler, mf_laplace._constraints,
+          mf_laplace._laplace_operator, material_property);
+  mfmg::Hierarchy<DVector> hierarchy(comm, evaluator, params, timer);
+
+  // We want to do 20 V-cycle iterations. The rhs of is zero.
+  // Use D(istributed)Vector because deal has its own Vector class
+  DVector residual(rhs);
+  unsigned int const n_cycles = 20;
+  std::vector<double> res(n_cycles + 1);
+
+  mf_laplace._laplace_operator.vmult(residual, solution);
+  residual.sadd(-1., 1., rhs);
+  auto const residual0_norm = residual.l2_norm();
+
+  std::cout << std::scientific;
+  res[0] = 1.0;
+  for (unsigned int i = 0; i < n_cycles; ++i)
+  {
+    hierarchy.vmult(solution, rhs);
+
+    mf_laplace._laplace_operator.vmult(residual, solution);
+    residual.sadd(-1., 1., rhs);
+    double rel_residual = residual.l2_norm() / residual0_norm;
+    res[i + 1] = rel_residual;
+  }
+
+  double const conv_rate = res[n_cycles] / res[n_cycles - 1];
+  pcout << "Convergence rate: " << std::fixed << std::setprecision(2)
```

Comment:
I guess we should aim for similar tolerances in the other tests as well...

Reply:
This is not a test. It's just an executable that we can use for profiling.